### PR TITLE
Add more e2e tests

### PR DESCRIPTION
**Pull request by fromanirh**
we increease the coverage of the topology updating tests, and we add the real rte-specific tests.

### CODE REVIEW

File: pkg/nrtupdater/nrtupdater.go [hunk 1]
```diff
@@ -15,6 +15,15 @@ import (
...
+const (
+	AnnotationRTEUpdate = "k8stopoawareschedwg/rte-update"
+)
+
+const (
+	RTEUpdatePeriodic = "periodic"
+	RTEUpdateReactive = "reactive"
+)
+
 var (
```
Comment (swatisehgal) [inline pkg/nrtupdater/nrtupdater.go:27]: There was feedback on the NFD PR to switch from using log to klog, looks like there is an inclination towards usingklog and we should consider doing that in RTE as well. Could be a completely separate PR.

Comment (fromanirh) [reply pkg/nrtupdater/nrtupdater.go:27]: ack, let's address indeed.

File: test/e2e/topology_updater.go [hunk 2]
```diff
@@ -41,35 +44,154 @@ import (
...
+			finalAllocRes := allocatableResourceListFromNodeResourceTopology(finalNodeTopo)
+			if len(initialAllocRes) == 0 || len(finalAllocRes) == 0 {
+				ginkgo.Fail(fmt.Sprintf("failed to find allocatable resources from node topology initial=%v final=%v", initialAllocRes, finalAllocRes))
+			}
+			zoneName, resName, cmp, ok := cmpAllocatableResources(initialAllocRes, finalAllocRes)
+			framework.Logf("zone=%q resource=%q cmp=%v ok=%v", zoneName, resName, cmp, ok)
+			if !ok {
+				ginkgo.Fail(fmt.Sprintf("failed to compare allocatable resources from node topology initial=%v final=%v", initialAllocRes, finalAllocRes))
+			}
+			// this is actually a workaround. The proper solution is to wait for ALL the container requesting exclusive resources to be gone
```
Comment (swatisehgal) [inline test/e2e/topology_updater.go:125]: My understanding was that at a given time for a particular test the pod above is running, is that not true? Based on that the `initialAllocRes` and `finalAllocRes` should be the same as the pod acquires CPUs from the shared pool. I am a little confused by this comment here that there might be other exclusive CPUs running?

File: test/e2e/topology_updater.go [hunk 3]
```diff
@@ -41,35 +44,154 @@ import (
...
+		})
+
+		ginkgo.It("it should account for containers requesting exclusive cpus", func() {
+			nodes, err := utils.FilterNodesWithEnoughCores(workerNodes, "1000m")
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			if len(nodes) < 1 {
+				ginkgo.Skip("not enough allocatable cores for this test")
+			}
+
+			ginkgo.By("getting the initial topology information")
```
Comment (swatisehgal) [inline test/e2e/topology_updater.go:138]: It seems a little unexpected to me that above we are saying that it should account for containers requesting exclusive cpus and then create a pod consuming resources from shared pool, I was expecting a pod requesting exclusive CPU as well. Given that we have a case above where we are only considering a pod in shared pool and resources are not decremented that means that in case we have two pods (one acquiring resources exclusively and another from shared pool)  if resources are decremented in finalAllocRes, it is safe to assume that it is happening because of the pod with exclusive CPUs? WDYT?

**Review by swatisehgal — COMMENTED** [review]
Thanks for the PR, I took a pass at it. Have a couple of comments/questions.

File: test/e2e/rte.go [hunk 4]
```diff
@@ -182,6 +183,61 @@ var _ = ginkgo.Describe("[RTE][InfraConsuming] Resource topology exporter", func
...
+		ginkgo.It("it should react to pod changes using the smart poller", func() {
+			nodes, err := utils.FilterNodesWithEnoughCores(workerNodes, "1000m")
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			if len(nodes) < 1 {
+				ginkgo.Skip("not enough allocatable cores for this test")
+			}
+
+			initialNodeTopo := getNodeTopology(topologyClient, topologyUpdaterNode.Name, namespace)
+			ginkgo.By("creating a pod consuming the shared pool")
+			sleeperPod := &v1.Pod{
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply test/e2e/rte.go:83]: I guess we can, with properly specialized functions. The intent of inlining the pod definiton is to keep all the context local to the test definition.

File: test/e2e/rte.go [hunk 5]
```diff
@@ -182,6 +183,61 @@ var _ = ginkgo.Describe("[RTE][InfraConsuming] Resource topology exporter", func
...
 
+		ginkgo.It("it should react to pod changes using the smart poller", func() {
+			nodes, err := utils.FilterNodesWithEnoughCores(workerNodes, "1000m")
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			if len(nodes) < 1 {
+				ginkgo.Skip("not enough allocatable cores for this test")
+			}
+
+			initialNodeTopo := getNodeTopology(topologyClient, topologyUpdaterNode.Name, namespace)
+			ginkgo.By("creating a pod consuming the shared pool")
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply test/e2e/rte.go:82]: ack

File: test/e2e/topology_updater.go [hunk 6]
```diff
@@ -41,35 +44,154 @@ import (
...
 		}
 
-		// intentionally get every single time
+		// intentionally done once
 		kubeletConfig, err = e2ekubelet.GetCurrentKubeletConfig(topologyUpdaterNode.Name, "", true)
 		gomega.Expect(err).NotTo(gomega.HaveOccurred())
 	})
 
 	ginkgo.Context("with cluster configured", func() {
+		ginkgo.It("it should not account for any cpus if a container doesn't request exclusive cpus", func() {
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply test/e2e/topology_updater.go:82]: good point. Adding.

File: test/e2e/topology_updater.go [hunk 7]
```diff
@@ -41,35 +44,154 @@ import (
...
+			ginkgo.By("getting the initial topology information")
+			initialNodeTopo := getNodeTopology(topologyClient, topologyUpdaterNode.Name, namespace)
+			ginkgo.By("creating a pod consuming the shared pool")
+			sleeperPod := &v1.Pod{
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply test/e2e/topology_updater.go:86]: a parametric `makePod` function didn't turn to be so helpful in the kube tests; because of the parametric nature the readability ended up a bit worse. But I guess we can try again.

File: test/e2e/topology_updater.go (hunk 2 above)
Comment (fromanirh) [reply test/e2e/topology_updater.go:125]: The comment is misplaced. Depending on the (random, by design) order on which ginkgo runs the tests, a test which exclusively allocates CPUs may run before. I didn't find yet a clean way to wait for this pod to be completely gone (hence, releasing the exclusively allocated CPUs) before to end the test, so *this* test can run with some leftovers hanging around, which makes the accounting harder.

I will improve and/or move the comment to capture this information.

File: test/e2e/topology_updater.go (hunk 3 above)
Comment (fromanirh) [reply test/e2e/topology_updater.go:138]: in this case the `ginkgo.By` statement (line 140) is a copy-paste mistake (my fault :) ), this test wants to get exclusively allocated CPUs indeed. Will fix.

File: test/e2e/utils/pod.go [hunk 8]
```diff
@@ -17,37 +17,84 @@ limitations under the License.
...
+			RestartPolicy: v1.RestartPolicyNever,
+			Containers: []v1.Container{
+				v1.Container{
+					Name:  "sleeper-gu-cnt",
+					Image: CentosImage,
+					// 1 hour (or >= 1h in general) is "forever" for our purposes
+					Command: []string{"/bin/sleep", "1h"},
+					Resources: v1.ResourceRequirements{
+						Limits: v1.ResourceList{
+							// we use 1 core because that's the minimal meaningful quantity
```
Comment (swatisehgal) [inline test/e2e/utils/pod.go:49]: How is this is a guaranteed pod if there is no resource request. For a pod to be guaranteed we have to make sure request=limit ?

File: test/e2e/topology_updater.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (swatisehgal) [reply test/e2e/topology_updater.go]: Oh okay, I saw them being heavily used when I worked on integration tests with SIG scheduling that is why recommended that. Thanks for making the change.

File: test/e2e/topology_updater.go (hunk 2 above)
Comment (swatisehgal) [reply test/e2e/topology_updater.go]: Thanks!

File: test/e2e/topology_updater.go [hunk 9]
```diff
@@ -79,27 +78,52 @@ var _ = ginkgo.Describe("[TopologyUpdater][InfraConsuming] Node topology updater
...
+			if len(initialAllocRes) == 0 || len(finalAllocRes) == 0 {
+				ginkgo.Fail(fmt.Sprintf("failed to find allocatable resources from node topology initial=%v final=%v", initialAllocRes, finalAllocRes))
+			}
+			zoneName, resName, cmp, ok := cmpAllocatableResources(initialAllocRes, finalAllocRes)
+			framework.Logf("zone=%q resource=%q cmp=%v ok=%v", zoneName, resName, cmp, ok)
+			if !ok {
+				ginkgo.Fail(fmt.Sprintf("failed to compare allocatable resources from node topology initial=%v final=%v", initialAllocRes, finalAllocRes))
 			}
 
+			// This is actually a workaround.
```
Comment (swatisehgal) [inline test/e2e/topology_updater.go:110]: This is definitely a TODO for us but appreciate you providing detailed explanation here.

Comment (fromanirh) [reply test/e2e/topology_updater.go:110]: ack, will mark it as such

File: test/e2e/utils/pod.go (hunk 8 above)
Comment (fromanirh) [reply test/e2e/utils/pod.go:49]: because https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#requests-and-limits - see note at the bottom of the paragraph.

Comment (swatisehgal) [reply test/e2e/utils/pod.go:49]: Ah nice! Today I learned :) Thanks
